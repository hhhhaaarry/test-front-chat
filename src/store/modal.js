import { defineStore } from "pinia";
export const useModalStore = defineStore("modal", {
  state: () => ({
    modal: false,
  }),
  actions: {
    openModal() {
      this.modal = true;
    },
    closeModal() {
      this.modal = false;
    },
    toggleModal() {
      this.modal = !this.modal;
    },
  },
  getters: {
    getModal() {
      return this.modal;
    },
  },
});
