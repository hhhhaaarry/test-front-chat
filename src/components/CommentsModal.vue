<template>
    <div
        class="modal container overflow-hidden min-w-0 sm:min-w-10 min-h-500 bg-slate-50 dark:bg-slate-900 fixed top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2 text-slate-700 dark:text-slate-200 shadow-lg">
        <header
            class="modal__header header flex flex-row justify-center w-full relative border-b-[1px] border-slate-400 dark:border-slate-700 p-3 px-6">
            <h2 class="modal__title header__title text-center text-lg font-bold">Comentarios</h2>
            <IconButton class="modal__close header__close absolute top-1/2 right-2 transform -translate-y-1/2"
                @click="modalStore.toggleModal">
                <XMarkIcon class="h-6 w-6" />
            </IconButton>
        </header>
        <main
            class="modal__main chat flex flex-col-reverse justify-between gap-3 p-3 px-6 custom-scrollbar overflow-auto max-h-96 ">
            <ChatItem v-for="(message, index) in sortedMessages" :key="index" :message="message" />
        </main>
        <div
            class="modal__input flex flex-row p-3 px-6 gap-2 leading-none text-sm border-t-[1px] border-slate-400 dark:border-slate-700">
            <input type="text"
                class="input input--rounded w-full rounded-3xl bg-slate-100 dark:bg-slate-800 border-[1px] border-slate-400 dark:border-slate-700 px-3"
                placeholder="Escribe un mensaje" v-model="newMessage" />
            <IconButton @click="sendMessage">
                <PaperAirplaneIcon class="h-6 w-6 fill-blue-500 dark:fill-blue-300" />
            </IconButton>
        </div>
    </div>
</template>

<script setup>
import { ref, onMounted } from 'vue'
import moment from 'moment'
import ChatItem from './ChatItem.vue'
import IconButton from './IconButton.vue'
import { XMarkIcon, PaperAirplaneIcon, ArrowDownTrayIcon } from '@heroicons/vue/24/solid'
import { useModalStore } from '../store/modal'

const modalStore = useModalStore()
const props = defineProps({
    messages: {
        type: Array,
        required: true
    }
})

let sortedMessages = ref([])

let newMessage = ref()

const sendMessage = () => {
    if (newMessage.value && newMessage.value.trim() !== '') {
        // Obtiene el avatar del Ãºltimo mensaje del usuario actual
        const avatar = sortedMessages.value.find(message => message.from === 'jid_1111')?.avatar || ''

        sortedMessages.value.unshift({
            type: 'text',
            from: 'jid_1111',
            text: newMessage.value, timestamp: moment().format(),
            avatar: avatar
        })
        newMessage.value = ''
    }
}

onMounted(() => {
    sortedMessages.value = props.messages.sort((a, b) => moment(b.timestamp) - moment(a.timestamp))
})
</script>
<style scoped>
.modal {
    max-width: calc(100% - 4rem);

    @media (min-width: 640px) {
        max-width: 600px;
    }
}
</style>
```

