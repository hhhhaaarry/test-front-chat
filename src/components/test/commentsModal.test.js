import { mount } from "@vue/test-utils";
import { useModalStore } from "../../store/modal";
import CommentsModal from "../CommentsModal.vue";
import { createPinia } from "pinia";

describe("CommentsModal.vue", () => {
  let wrapper;

  beforeEach(() => {
    const pinia = createPinia();
    const modalStore = useModalStore(pinia); // Pasa la instancia de Pinia a useModalStore

    wrapper = mount(CommentsModal, {
      propsData: {
        messages: [
          {
            type: "text",
            from: "jid_1111",
            text: "Hola",
            timestamp: "2022-01-01T00:00:00Z",
            avatar: "avatar.png",
          },
        ],
      },
      global: {
        plugins: [pinia], // Instala Pinia en el contexto de la prueba
        provide: {
          modalStore: {
            toggleModal: modalStore.toggleModal,
            getModal: modalStore.getModal,
            modal: modalStore.modal,
          },
        },
      },
    });
  });

  // Verifica que el componente se renderiza correctamente
  it("se renderiza correctamente", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  // Verifica que el modal se abre cuando pulsas el botón
  it("se abre el modal cuando pulsas el botón", async () => {
    const button = wrapper.find("button");
    await button.trigger("click");
    expect(wrapper.vm.modalStore.modal).toBe(true);
  });

  // Verifica que el modal se cierra cuando pulsas el botón
  it("se cierra el modal cuando pulsas el botón", async () => {
    const button = wrapper.find("button");
    await button.trigger("click");
    expect(wrapper.vm.modalStore.modal).toBe(true);
    await button.trigger("click");
    expect(wrapper.vm.modalStore.modal).toBe(false);
  });
  // Verifica que se envía un mensaje cuando se hace clic en el botón de enviar
  it("envía un mensaje al hacer clic en el botón de enviar", async () => {
    const input = wrapper.find(".modal__input input");
    const sendButton = wrapper.find(".modal__input button");
    await input.setValue("Hola mundo");
    await sendButton.trigger("click");
    expect(wrapper.vm.sortedMessages[0].text).toBe("Hola mundo");
  });
  // Verifica que los mensajes se ordenan por timestamp
  it("ordena los mensajes por timestamp", () => {
    const messages = wrapper.vm.sortedMessages;
    for (let i = 0; i < messages.length - 1; i++) {
      expect(messages[i].timestamp).toBeGreaterThanOrEqual(
        messages[i + 1].timestamp
      );
    }
  });
  // Verifica que el input se limpia después de enviar un mensaje
  it("limpia el input después de enviar un mensaje", async () => {
    const input = wrapper.find(".modal__input input");
    const sendButton = wrapper.find(".modal__input button");

    await input.setValue("Hola mundo");
    await sendButton.trigger("click");

    expect(input.element.value).toBe("");
  });
});
