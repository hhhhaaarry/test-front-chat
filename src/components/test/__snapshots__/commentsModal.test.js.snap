// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommentsModal.vue > se renderiza correctamente 1`] = `
"<div class=\\"modal container max-w-lg overflow-hidden min-w-10 min-h-500 bg-slate-50 dark:bg-slate-900 fixed top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2 text-slate-700 dark:text-slate-200 shadow-lg\\">
  <header class=\\"modal__header header flex flex-row justify-center w-full relative border-b-[1px] border-slate-400 dark:border-slate-700 p-3 px-6\\">
    <h2 class=\\"modal__title header__title text-center text-lg font-bold\\">Comentarios</h2><button class=\\"btn btn--icon bg-slate-50 dark:bg-gray-800 p-2 flex justify-center items-center hover:border-blue-500 dark:hover:border-gray-200 modal__close header__close absolute top-1/2 right-2 transform -translate-y-1/2\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" fill=\\"currentColor\\" aria-hidden=\\"true\\" class=\\"h-6 w-6\\">
        <path fill-rule=\\"evenodd\\" d=\\"M5.47 5.47a.75.75 0 011.06 0L12 10.94l5.47-5.47a.75.75 0 111.06 1.06L13.06 12l5.47 5.47a.75.75 0 11-1.06 1.06L12 13.06l-5.47 5.47a.75.75 0 01-1.06-1.06L10.94 12 5.47 6.53a.75.75 0 010-1.06z\\" clip-rule=\\"evenodd\\"></path>
      </svg></button>
  </header>
  <main class=\\"modal__main chat flex flex-col-reverse justify-between gap-3 p-3 px-6 custom-scrollbar overflow-auto max-h-96\\">
    <div class=\\"chat-item rounded-sm p-2 max-w-full w-full text-left flex flex-row justify-end items-center gap-3 text-slate-700 dark:text-white chat-item--sent\\">
      <div class=\\"chat-item__blob p-3 px-6 relative text-base leading-tight flex flex-col gap-1 max-w-[85%] bg-blue-600 text-white\\">
        <div class=\\"chat-item__text\\">Hola</div>
        <div class=\\"chat-item__timestamp\\"><span class=\\"text-xs dark:text-gray-300\\">01 ene. 2022 / 01: 00</span></div>
      </div>
      <div class=\\"chat-item__avatar h-6 w-6 flex shrink-0 rounded-2xl overflow-hidden\\"><img class=\\"w-full h-full object-cover object-center\\" src=\\"avatar.png\\" alt=\\"\\"></div>
    </div>
  </main>
  <div class=\\"modal__input flex flex-row p-3 px-6 gap-2 leading-none text-sm border-t-[1px] border-slate-400 dark:border-slate-700\\"><input type=\\"text\\" class=\\"input input--rounded w-full rounded-3xl bg-slate-100 dark:bg-slate-800 border-[1px] border-slate-400 dark:border-slate-700 px-3\\" placeholder=\\"Escribe un mensaje\\"><button class=\\"btn btn--icon bg-slate-50 dark:bg-gray-800 p-2 flex justify-center items-center hover:border-blue-500 dark:hover:border-gray-200\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" fill=\\"currentColor\\" aria-hidden=\\"true\\" class=\\"h-6 w-6 fill-blue-500 dark:fill-blue-300\\">
        <path d=\\"M3.478 2.405a.75.75 0 00-.926.94l2.432 7.905H13.5a.75.75 0 010 1.5H4.984l-2.432 7.905a.75.75 0 00.926.94 60.519 60.519 0 0018.445-8.986.75.75 0 000-1.218A60.517 60.517 0 003.478 2.405z\\"></path>
      </svg></button></div>
</div>"
`;
